### PR TITLE
Crawl the source map files

### PR DESCRIPTION
**Pull request by ShubhamRasal**
- Crawl the source map files
- Use body parser to parse the sourceContents

<details><summary>Example 1</summary> 

```console
❯ ./katana -u https://bitbucket.org -jc -ns -v

   __        __                
  / /_____ _/ /____ ____  ___ _
 /  '_/ _  / __/ _  / _ \/ _  /
/_/\_\\_,_/\__/\_,_/_//_/\_,_/							 

		projectdiscovery.io

[INF] Current katana version v1.0.2-dev (dev)
[INF] Started standard crawling for => https://bitbucket.org

[html] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js
[html] [GET] https://www.facebook.com/djeihiulvfkgcccw.pkg.js
[js] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js.map
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/common/tracking
[map] [GET] https://wac-cdn.atlassian.com/air-datepicker.js
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/common/analytics
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/utils/general
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/app.js
[map] [GET] https://wac-cdn.atlassian.com/utils/localized-pricing
[map] [GET] https://wac-cdn.atlassian.com/utils/url
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/utils/localized-pricing
[map] [GET] https://wac-cdn.atlassian.com/utils/storage
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/utils/url
[map] [GET] https://wac-cdn.atlassian.com/utils/string

```

</details>

<details><summary>Example 2</summary> 

```console
❯ ./katana -u https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js -jc -ns  -v

   __        __                
  / /_____ _/ /____ ____  ___ _
 /  '_/ _  / __/ _  / _ \/ _  /
/_/\_\\_,_/\__/\_,_/_//_/\_,_/							 

		projectdiscovery.io

[INF] Current katana version v1.0.2-dev (dev)
[INF] Started standard crawling for => https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js
[GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js
[html] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js
[js] [GET] https://api-private.atlassian.com/available-sites
[js] [GET] https://api-private.atlassian.com/me
[js] [GET] https://api-private.stg.atlassian.com/available-sites
[js] [GET] https://api-private.stg.atlassian.com/me
[js] [GET] http://www.w3.org/TR/SVG11/feature
[js] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/bitbucket.js.map
[header] [GET] https://www.w3.org/TR/SVG11/feature
[js] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/underscore.js
[js] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/air-datepicker.js
[js] [GET] https://wac-cdn.atlassian.com/login/select-account?continue=
[html] [GET] http://www.w3.org/1999/xlink
[map] [GET] https://wac-cdn.atlassian.com/common/tracking/platforms
[html] [GET] https://wac-cdn.atlassian.com/master/4501/assets/build/js/ad-hoc/bitbucket/air-datepicker.js
[html] [GET] https://www.w3.org/expanders.js
[html] [GET] https://wac-cdn.atlassian.com/master/4501/assets/build/js/ad-hoc/bitbucket/underscore.js
[js] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/select2.js
[html] [GET] https://wac-cdn.atlassian.com/dam/jcr
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/utils/history
[script] [GET] https://atl-global.atlassian.com/js/atl-global.min.js
[html] [GET] https://wac-cdn.atlassian.com/master/4501/assets/build/js/ad-hoc/bitbucket/select2.js
[js] [GET] https://www.atlassian.com/hamlet/1.0/auth/xsrf/html?domain=atlassian.com
[script] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/head.js?cdnVersion=971
[html] [GET] https://confluence.atlassian.com/display/ALLDOC/Atlassian
[script] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/jquery.js?cdnVersion=971
[js] [GET] https://www.atlassian.com/gateway/api
[a] [GET] https://www.facebook.com/Atlassian
[a] [GET] https://wac-cdn.atlassian.com/legal/impressum
[a] [GET] https://wac-cdn.atlassian.com/legal/cloud-terms-of-service
[script] [GET] https://wac-cdn.atlassian.com/login/bower_components/
[script] [GET] https://wac-cdn.atlassian.com/assets/
[a] [GET] https://wac-cdn.atlassian.com/company/contact
[a] [GET] https://wac-cdn.atlassian.com/legal/privacy-policy
[a] [GET] https://wac-cdn.atlassian.com/company/events
[link] [GET] https://wac-cdn.atlassian.com
[a] [GET] https://wac-cdn.atlassian.com/resources
[a] [GET] https://support.atlassian.com/contact/
[a] [GET] https://wac-cdn.atlassian.com/developers
[a] [GET] https://confluence.atlassian.com/kb
[link] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/css/wpl-main.css?cdnVersion=971
[a] [GET] https://www.atlassian.com/nl/404
[map] [GET] https://wac-cdn.atlassian.com/common/tracking
[a] [GET] https://www.atlassian.com/purchase/
[map] [GET] https://wac-cdn.atlassian.com/common/analytics
[a] [GET] https://www.atlassian.com/es/404
[a] [GET] https://www.atlassian.com/it/404
[a] [GET] https://www.atlassian.com/fr/404
[a] [GET] https://www.atlassian.com/404
[a] [GET] https://www.atlassian.com/de/404
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/common/tracking/platforms
[a] [GET] https://www.atlassian.com/zh/404
[map] [GET] https://wac-cdn.atlassian.com/underscore.js
[map] [GET] https://wac-cdn.atlassian.com/app.js
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/common/tracking
[map] [GET] https://wac-cdn.atlassian.com/air-datepicker.js
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/common/analytics
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/utils/general
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/app.js
[map] [GET] https://wac-cdn.atlassian.com/utils/localized-pricing
[map] [GET] https://wac-cdn.atlassian.com/utils/url
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/utils/localized-pricing
[map] [GET] https://wac-cdn.atlassian.com/utils/storage
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/utils/url
[map] [GET] https://wac-cdn.atlassian.com/utils/string
[map] [GET] https://wac-cdn.atlassian.com/OS%5C/2/
[map] [GET] https://wac-cdn.atlassian.com/BeOS/
[map] [GET] https://wac-cdn.atlassian.com/UNIX/
[map] [GET] https://wac-cdn.atlassian.com/QNX/
[map] [GET] https://wac-cdn.atlassian.com/SunOS/
[map] [GET] https://wac-cdn.atlassian.com/utils/history
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/utils/storage
[map] [GET] https://wac-cdn.atlassian.com/OpenBSD/
[map] [GET] https://wac-cdn.atlassian.com/Android/
[map] [GET] https://wac-cdn.atlassian.com/utils/tracking
[map] [GET] https://wac-cdn.atlassian.com/Win16/
[map] [GET] https://wac-cdn.atlassian.com/utils/general
[map] [GET] https://wac-cdn.atlassian.com/static/master/4501/assets/build/js/ad-hoc/bitbucket/utils/string
[map] [GET] https://wac-cdn.atlassian.com/utils/get-script
[map] [GET] https://wac-cdn.atlassian.com/utils/api
[map] [GET] https://wac-cdn.atlassian.com/utils/browser
```
</details>

### CODE REVIEW

**Review by Mzack9999 — APPROVED** [review]
lgtm! - Question: By unmarshaling the json, we might detect more decoded endpoints, as many characters are escaped in the marshaled json and might not match the regexes?

File: pkg/engine/parser/parser.go [hunk 1]
```diff
@@ -638,16 +639,48 @@ func scriptJSFileRegexParser(resp *navigation.Response) (navigationRequests []*n
...
 	endpoints := utils.ExtractRelativeEndpoints(string(resp.Body))
 	for _, item := range endpoints {
 		navigationRequests = append(navigationRequests, navigation.NewNavigationRequestURLFromResponse(item, resp.Resp.Request.URL.String(), "js", "regex", resp))
 	}
+
+	if strings.Contains(string(resp.Body), "//# sourceMappingURL=") {
+		url := resp.Resp.Request.URL
+		url.RawQuery = ""
+		navigationRequests = append(navigationRequests, navigation.NewNavigationRequestURLFromResponse(url.String()+".map", resp.Resp.Request.URL.String(), "js", "eval", resp))
+	}
```
Comment (tarunKoyalwar) [inline pkg/engine/parser/parser.go:651]: @ShubhamRasal , not sure if there is a need to remove any existing parameters . we can clone any url struct using 
```go
	temp := urlutil.URL{URL: resp.Resp.Request.URL}
	cloned := temp.Clone()
```
and we can add to cloned.Path directly

File: pkg/types/sourcemap.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package types
+
+// SourceMap is a struct for source map
+type SoruceMap struct {
```
Comment (tarunKoyalwar) [inline pkg/types/sourcemap.go:4]: typo `SourceMap`

**Review by tarunKoyalwar — COMMENTED** [review]
lgtm !

suggesting minor changes

**Comment by tarunKoyalwar** [comment]
@Mzack9999 , a sample sourcemap file is available at below url 
```console
curl https://wac-cdn.atlassian.com/static/master/4466/assets/build/js/chunks/9b24561751aa1c.js.map | jq .
```
as you can tell running Endpoint regex and other regex on sourcemap content already generates lot of false positives this is due to 
- `filePaths` like `./utils` are considered as relative endpoints
- `third party` node_modules
- `mappings` & variable names might be considered as endpoint by regex

by unmarshalling and removing `node_modules` we are actually trying to reduce false positives .

ideal solution would be to also add a new regex since it is quite rare to find a url like `https://somethig.com/../abc` but we will definitely see relative js/ts import statements like `import ./utils`

**Review by ehsandeep — COMMENTED** [review]
```console
./katana -u https://wac-cdn.atlassian.com/static/master/4466/assets/build/js/chunks/9b24561751aa1c.js.map -jc -ns

   __        __                
  / /_____ _/ /____ ____  ___ _
 /  '_/ _  / __/ _  / _ \/ _  /
/_/\_\\_,_/\__/\_,_/_//_/\_,_/							 

		projectdiscovery.io

[INF] Current katana version v1.0.2-dev (dev)
[INF] Started standard crawling for => https://wac-cdn.atlassian.com/static/master/4466/assets/build/js/chunks/9b24561751aa1c.js.map
https://wac-cdn.atlassian.com/static/master/4466/assets/build/js/chunks/9b24561751aa1c.js.map
```

**Comment by ShubhamRasal** [comment]
image

**Review by ehsandeep — APPROVED** [review]

**Review by ehsandeep — CHANGES_REQUESTED** [review]
```console
./katana -u https://bitbucket.org -jc -ns -v -proxy http://127.0.0.1:8080

   __        __                
  / /_____ _/ /____ ____  ___ _
 /  '_/ _  / __/ _  / _ \/ _  /
/_/\_\\_,_/\__/\_,_/_//_/\_,_/							 

		projectdiscovery.io

[INF] Current katana version v1.0.4 (latest)
[INF] Started standard crawling for => https://bitbucket.org
[GET] https://bitbucket.org
[html] [GET] http://www.w3.org/2000/svg
[script] [GET] https://atl-global.atlassian.com/js/atl-global.min.js
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x10390b1fc]

goroutine 31 [running]:
github.com/projectdiscovery/utils/url.(*OrderedParams).Clone(0x0)
	/Users/geekboy/go/pkg/mod/github.com/projectdiscovery/utils@v0.0.60/url/orderedparams.go:157 +0x6c
github.com/projectdiscovery/utils/url.(*URL).Clone(0x1400128fd50)
	/Users/geekboy/go/pkg/mod/github.com/projectdiscovery/utils@v0.0.60/url/url.go:90 +0x204
github.com/projectdiscovery/katana/pkg/engine/parser.scriptJSFileRegexParser(0x140012a2790)
	/Users/geekboy/Github/katana/pkg/engine/parser/parser.go:674 +0x29c
github.com/projectdiscovery/katana/pkg/engine/parser.ParseResponse(0x140012a2790)
	/Users/geekboy/Github/katana/pkg/engine/parser/parser.go:101 +0x244
github.com/projectdiscovery/katana/pkg/engine/common.(*Shared).Do.func1()
	/Users/geekboy/Github/katana/pkg/engine/common/base.go:246 +0x174
created by github.com/projectdiscovery/katana/pkg/engine/common.(*Shared).Do in goroutine 61
	/Users/geekboy/Github/katana/pkg/engine/common/base.go:214 +0x250
```

**Comment by dogancanbakir** [comment]
Closing this due to lost context. We'll review the issue again.

**Pull request closed without merging**